src/main/java/bio/terra/service/dataset/DatasetService.java

```diff
@@ -353,6 +355,91 @@ public String addDatasetAssetSpecifications(
+        dataset.getTables().stream()
+            .filter(datasetTable -> datasetTable.getName().equals(assetModel.getRootTable()))
+            .findFirst()
+            .orElseThrow(
+                () ->
+                    new InvalidAssetException(
+                        "Root table " + assetModel.getRootTable() + " does not exist in dataset."));
+    // Validate Root Column
+    if (!rootTable.getColumns().stream()
+        .map(c -> c.getName())
+        .toList()
+        .contains(assetModel.getRootColumn())) {
+      throw new InvalidAssetException(
+          "Root column "
+              + assetModel.getRootColumn()
+              + " does not exist in table "
+              + rootTable.getName());
+    }
+    // Validate Tables
+    assetModel
+        .getTables()
+        .forEach(
+            assetTable -> {
+              DatasetTable currentTable =
+                  dataset.getTables().stream()
+                      .filter(datasetTable -> datasetTable.getName().equals(assetTable.getName()))
+                      .findFirst()
+                      .orElseThrow(
+                          () ->
+                              new InvalidAssetException(
+                                  "Table " + assetTable.getName() + " does not exist in dataset."));
+              List<String> datasetTableColumnNames =
+                  currentTable.getColumns().stream().map(c -> c.getName()).toList();
+              assetTable
+                  .getColumns()
+                  .forEach(
+                      assetColumn -> {
+                        if (!datasetTableColumnNames.contains(assetColumn)) {
+                          throw new InvalidAssetException(
+                              "Column "
+                                  + assetColumn
+                                  + " does not exist in table "
+                                  + assetTable.getName());
+                        }
+                      });
+            });
+
+    // Follow should reference an existing relationship as defined in the original dataset create
+    // query
+    assetModel
+        .getFollow()
+        .forEach(
+            follow -> {
+              if (!dataset.getRelationships().stream()
+                  .map(r -> r.getName())
+                  .toList()
+                  .contains(follow)) {
+                throw new InvalidAssetException(
+                    "Relationship specified in follow list '"
+                        + follow
+                        + "' does not exist in dataset's list of relationships");
+              }
+            });
+  }
+
+  public AssetSpecification getNewAssetSpec(Dataset dataset, AssetModel assetModel) {
+    List<DatasetTable> datasetTables = dataset.getTables();
+    Map<String, Relationship> relationshipMap = new HashMap<>();
+    Map<String, DatasetTable> tablesMap = new HashMap<>();
+
+    datasetTables.forEach(datasetTable -> tablesMap.put(datasetTable.getName(), datasetTable));
```

Comment:
You can write this as streaming
```suggestion
    Map<String, DatasetTable> tablesMap =
        datasetTables.forEach(datasetTable -> tablesMap.put(datasetTable.getName(), datasetTable));
```

Reply:
I ended up switching this to a more direct mapping into the map. Hopefully, that makes sense. 
It does make me a little nervous to change up this code b/c there aren't any unit tests for this older code (I just moved it, didn't previously modify the code). So, an extra look to make sure I didn't change functionality would be appreciated!